all: clean up and properly abstract database accesses

karalabe opened the pull request:
This is a preparatory PR for the chain freezer https://github.com/ethereum/go-ethereum/pull/17814 and historical state pruner https://github.com/ethereum/go-ethereum/pull/18418 works.

Our database package was kind of working until now, but the interfaces were convoluted a bit, many many packages redeclared some random portion of it, and certain features were not implemented by all backends (iterators). This PR attempts to properly clean up all the interfaces related to databases with proper backend implementations being at feature parity with each other.

The PR itself is a NOOP. No extra functionality or features were included whatsoever, only existing code refactored. The PR was separated out from the chain freezer (see last commit), so it might happen that some part may look like unused, but actually everything is needed and fits nicely into a coherent design.

An important design choice behind this work is that I introdced a `database` package, which just defines the interfaces, but contains no implementation. This is essential to allow all our code to use the same interfaces and not have to redeclare them over and over and over again. Adding implementation here results in circular dependencies.

All actual implementations live in `internal/keyvalue` (the low level KeyValueStores) and `core/rawdb` is responsible for constructing actual `database.Database` instances, which in this PR are essentially the same as the key-value stores, but the freezer will add the extra functionality in a followup PR.

holiman commented on cmd/geth/chaincmd.go:
```diff
@@ -411,12 +409,12 @@ func copyDb(ctx *cli.Context) error {
 
 	// Compact the entire database to remove any sync overhead
-	start = time.Now()
+	/*start = time.Now()
```
So explicit compaction is now gone?

karalabe replied:
Ah, yes. Compaction requires casting it down to `leveldb.Database`. I could add compactions to the database interface.

holiman replied:
We have this:
```
	NoCompactionFlag = cli.BoolFlag{
		Name:  "nocompaction",
		Usage: "Disables db compaction after import",
	}
```
I make use of it in hive to prevent compaction. If you have disabled explicit compaction after import anyway, I guess you could remove that flag? I didn't actually see any change in that area, but I assume you must have done _something_ on that part too?  

But perhaps it would be better to add it to the interface.

karalabe replied:
Please see my last commit. I've added compaction support back.

holiman commented on internal/keyvalue/memorydb.go:
```diff
@@ -0,0 +1,295 @@
+	var (
+		pr     = string(prefix)
+		keys   = []string{"FAKE KEY TO SIMPLIFY NEXT"}
+		values = [][]byte{[]byte("FAKE VAL TO SIMPLIFY NEXT")}
```
This looks very much like code left in from a debugging session. If that's not the case, perhaps some more docs could be added?

karalabe replied:
Iteration works by initially placing the iterator *before* the first item, requiring a call to `Next` to actually advance to the first one. This special case can either be done explicitly within the iterator code, or in my case I made a hack whereby there's actually a 0-th dummy element that the iterator is positioned on when created.

karalabe replied:
An alternative code would be

```go
diff --git a/internal/keyvalue/memorydb.go b/internal/keyvalue/memorydb.go
index 453eba0a0..acee1dcc5 100644
--- a/internal/keyvalue/memorydb.go
+++ b/internal/keyvalue/memorydb.go
@@ -142,8 +142,8 @@ func (db *MemoryDatabase) NewIteratorWithPrefix(prefix []byte) database.Iterator
 
        var (
                pr     = string(prefix)
-               keys   = []string{"FAKE KEY TO SIMPLIFY NEXT"}
-               values = [][]byte{[]byte("FAKE VAL TO SIMPLIFY NEXT")}
+               keys   = make([]string, 0, len(db.db))
+               values = make([][]byte, 0, len(db.db))
        )
        // Collect the keys from the memory database corresponding to the given prefix
        for key := range db.db {
@@ -152,8 +152,8 @@ func (db *MemoryDatabase) NewIteratorWithPrefix(prefix []byte) database.Iterator
                }
        }
        // Sort the items and retrieve the associated values
-       sort.Strings(keys[1:])
-       for _, key := range keys[1:] {
+       sort.Strings(keys)
+       for _, key := range keys {
                values = append(values, db.db[key])
        }
        return &memoryIterator{
@@ -178,22 +178,6 @@ func (db *MemoryDatabase) Len() int {
        return len(db.db)
 }
 
-/*
-// Keys retrieves a copy of all the keys present in the database.
-//
-// Note, this method is only used for testing (i.e. not public in general) and
-// does not have explicit checks for closed-ness to allow simpler testing code.
-func (db *MemoryDatabase) Keys() [][]byte {
-       db.lock.RLock()
-       defer db.lock.RUnlock()
-
-       keys := [][]byte{}
-       for key := range db.db {
-               keys = append(keys, []byte(key))
-       }
-       return keys
-}*/
-
 // keyvalue is a key-value tuple tagged with a deletion field to allow creating
 // memory-database write batches.
 type keyvalue struct {
@@ -254,6 +238,7 @@ func (b *memoryBatch) Reset() {
 // key value store. Internally it is a deep copy of the entire iterated state,
 // sorted by keys.
 type memoryIterator struct {
+       inited bool
        keys   []string
        values [][]byte
 }
@@ -261,6 +246,12 @@ type memoryIterator struct {
 // Next moves the iterator to the next key/value pair. It returns whether the
 // iterator is exhausted.
 func (it *memoryIterator) Next() bool {
+       // If the iterator was not yet initialized, do it now
+       if !it.inited {
+               it.inited = true
+               return len(it.keys) > 0
+       }
+       // Iterator already initialize, advance it
        if len(it.keys) > 0 {
                it.keys = it.keys[1:]
                it.values = it.values[1:]
```

Perhaps it's better to keep it clean vs. dummy values.

holiman replied:
I don't really care, just wanted to check that it wasn't just some stale leftover

karalabe replied:
Please see my last commit. I've swapped out the dummy items with special case handling.

Matthalp commented on core/rawdb/database.go:
```diff
@@ -0,0 +1,51 @@
+// Copyright 2018 The go-ethereum Authors
```
Could you help me understand why this code was moved to `rawdb`? It seems like `database` was a good fit for this. IMHO `rawdb` provides low-level utility methods for writing `core` blockchain abstractions whereas `database` provides more low-level database abstractions and some starter implementations (e.g. the in-memory and LevelDB ones). This is especially true because the return types for these methods are from `database`.

karalabe replied:
Creating the low level database needs to be able to inject the freezer. The freezer needs access to higher level types (blocks, etc), so it needs rawdb. Putting it into `database` introduces a dependency loop.

Matthalp commented on core/rawdb/table.go:
```diff
@@ -0,0 +1,150 @@
+// Copyright 2018 The go-ethereum Authors
```
Similar points as in `database.go` about whether this is the right package for this to live in. However, this is more complicated as `Table` contains the prefix for where particular data is stored. This makes have a clear separation of abstractions more difficult.

Matthalp commented on database/batch.go:
```diff
@@ -14,39 +14,31 @@
-// Code using batches should try to add this much data to the batch.
-// The value was determined empirically.
+// IdealBatchSize defines the size of the data batches should ideally add in one
+// write.
```
Out of curiousity: was this number determined to be system-agnostic? It might be worth adding that in the comment to help provide clarity.

Matthalp commented on database/database.go:
```diff
@@ -0,0 +1,85 @@
+
+// KeyValueStore contains all the methods required to allow handling different
+// key-value data stores backing the high level database.
+type KeyValueStore interface {
```
Is there a reason to have separate `KeyValueStore` and `Database` interfaces? It feels redundant now since both have the exact same interface, but maybe there is a long-term plan for them to deviate?

karalabe replied:
The freezer deviates it, @fjl just wanted to have a separate PR with the rework to keep the two easier to review.

Matthalp commented on database/database.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+// Database contains all the methods required by the high level database to not
+// only access
```
Nit: punctuate.

karalabe replied:
Thx

Matthalp commented on database/iterator.go:
```diff
@@ -0,0 +1,57 @@
+
+package database
+
+// Iterator iterates over a databases key/value pairs in key order.
```
Nit: database's

karalabe replied:
Thx

Matthalp commented on database/iterator.go:
```diff
@@ -0,0 +1,57 @@
+
+package database
+
+// Iterator iterates over a databases key/value pairs in key order.
```
Nit: "ascending key order"

karalabe replied:
Thx

Matthalp commented on database/iterator.go:
```diff
@@ -0,0 +1,57 @@
+// Iterator iterates over a databases key/value pairs in key order.
+//
+// When it encounters an error any seek will return false and will yield no key/
+// value pairs. The error can be queried by calling the Error method. Calling
```
Should there be an `Error() error` method on this interface?

karalabe replied:
Thx, added

Matthalp commented on internal/keyvalue/leveldb.go:
```diff
@@ -26,23 +26,38 @@ import (
+// LeveldbDatabase is a persistent key-value store. Apart from basic data storage
+// functionality it also supports batch writes and iterating over the keyspace in
+// binary-alphabetical order.
+type LeveldbDatabase struct {
```
Nit: This is true across the codebase but figured I would raise it here first. Idiomatic golang capitalizes abbreviation names. So the "db" in `levelDb*` and `Leveldb*` should be `levelDB*` and `LevelDB*`. This is true with `chainDb` and `stateDB` variable names as well as names that contain subprotocols in it `Les*` should be  `LES*`.

https://github.com/golang/go/wiki/CodeReviewComments#initialisms

karalabe replied:
Fair enough, fixed.

Matthalp commented on internal/keyvalue/leveldb.go:
```diff
@@ -60,17 +75,16 @@ type LDBDatabase struct {
-func NewLDBDatabase(file string, cache int, handles int) (*LDBDatabase, error) {
-	logger := log.New("database", file)
-
+// NewLeveldbDatabase returns a wrapped LevelDB object.
```
It may be worth adding the purpose of the variable names in this method do to the comment. For example, it was not clear that `namespace` was for metrics until looking at the code..

karalabe replied:
Agreed. Done.

Matthalp commented on internal/keyvalue/leveldb.go:
```diff
@@ -83,102 +97,116 @@ func NewLDBDatabase(file string, cache int, handles int) (*LDBDatabase, error) {
-func (db *LDBDatabase) Path() string {
-	return db.fn
-}
+// Close stops th emetrics collection, flushes any pending data to disk and closes
```
s/the emetrics/the metrics/

karalabe replied:
thx

Matthalp commented on internal/keyvalue/leveldb.go:
```diff
@@ -83,102 +97,116 @@ func NewLDBDatabase(file string, cache int, handles int) (*LDBDatabase, error) {
-	return db.fn
-}
+// Close stops th emetrics collection, flushes any pending data to disk and closes
+// all io access to the underlying key-value store.
```
s/access/accesses/

karalabe replied:
thx

Matthalp commented on internal/keyvalue/leveldb.go:
```diff
@@ -83,102 +97,116 @@ func NewLDBDatabase(file string, cache int, handles int) (*LDBDatabase, error) {
 
-// Get returns the given key if it's present.
-func (db *LDBDatabase) Get(key []byte) ([]byte, error) {
+// Get retrieves the given key if it's present in the key-value store.
```
Nit: move "in the key-value store" immediately after "the given key"

karalabe replied:
I think it's fine, "if it's present in the key-value store"

Matthalp commented on internal/keyvalue/leveldb.go:
```diff
@@ -83,102 +97,116 @@ func NewLDBDatabase(file string, cache int, handles int) (*LDBDatabase, error) {
 // the metrics subsystem.
 //
-// This is how a stats table look like (currently):
+// This is how a stats leveldbTable look like (currently):
```
Nit: This would read better as `// This is what a LevelDB stats table currently looks like:`

karalabe replied:
I think this suffered a find/replace. Fixed it either way.

Matthalp commented on internal/keyvalue/leveldb.go:
```diff
@@ -221,19 +249,19 @@ func (db *LDBDatabase) meter(refresh time.Duration) {
 			continue
 		}
-		// Find the compaction table, skip the header
+		// Find the compaction leveldbTable, skip the header
```
Was the change from "table" to "leveldbTable" from an automated code change?

karalabe replied:
Probably

Matthalp commented on internal/keyvalue/memorydb.go:
```diff
@@ -0,0 +1,291 @@
+
+// memoryIterator can walk over the (potentially partial) keyspace of a memory
+// key value store. Internally it is a deep copy of the entire iterated state,
+// sorted by keys.
```
I think this comment would be more clear as "Internally it keeps a copy of the remaining state to be iterated".`keys` and `values` decreases in size by one each time `Next` is called and from what I can tell the copy is shallow.

karalabe replied:
The slices don't decrease in memory size, they still address the same underlying buffer, just the starting offset is bumped. Deep copy as in you can modify the database while the iterator is running.

Matthalp replied:
Right. What I meant to say was that the length of the slices being referenced are reducing by one each time (and not it's backing array). I also misunderstood that you meant the deep copy was from the database. Make sense.

karalabe commented on core/rawdb/accessors_chain.go:
```diff
@@ -233,15 +258,21 @@ func WriteBody(db DatabaseWriter, hash common.Hash, number uint64, body *types.B
-// ReadTd retrieves a block's total difficulty corresponding to the hash.
-func ReadTd(db DatabaseReader, hash common.Hash, number uint64) *big.Int {
+// ReadTdRLP retrieves a block's total difficulty corresponding to the hash in RLP encoding.
+func ReadTdRLP(db database.Reader, hash common.Hash, number uint64) rlp.RawValue {
```
Mostly to remain consistent with the other APIs. There's a ReadXyzRLP method for almost everything else, figured we might as well make this available too.

Matthalp replied:
I don't have a strong opinion here, but I generally like to add things when they are needed.

karalabe commented on core/rawdb/accessors_chain.go:
```diff
@@ -37,21 +38,39 @@ func ReadCanonicalHash(db DatabaseReader, number uint64) common.Hash {
+//
+// This method is a helper for the chain reader. It should never be exposed to the
+// outside world.
+func readAllHashes(db database.Iteratorer, number uint64) []common.Hash {
```
It was part of the chain freezer code, we can delete it from this pr.

karalabe commented on core/rawdb/accessors_chain.go:
```diff
@@ -265,25 +296,31 @@ func WriteTd(db DatabaseWriter, hash common.Hash, number uint64, td *big.Int) {
 }
 
+// ReadReceiptsRLP retrieves all the transaction receipts belonging to a block in RLP encoding.
+func ReadReceiptsRLP(db database.Reader, hash common.Hash, number uint64) rlp.RawValue {
```
Same as ReadTdRLP, makes the API most consistent.

karalabe commented on core/rawdb/database.go:
```diff
@@ -0,0 +1,51 @@
+// NewDatabase creates a high level database on top of a given key-value data
+// store without a freezer moving immutable chain segments into cold storage.
+func NewDatabase(db database.KeyValueStore) database.Database {
+	return db
```
I've just left this is so we don't diverge from the chain freezer code, which imho is still desired.

karalabe commented:
@Matthalp PTAL

Matthalp commented:
Thanks for addressing my comments in the code. All that's missing are some tests for the new `Iterator` code and the it will LGTM.

karalabe commented:
@Matthalp Which iterator code are you referring to? The crux of the iterator is implemented by leveldb.

Matthalp commented:
@karalabe I was referring to the new `database.Iterator` code. I would consider adding some tests as a sanity check to the LevelDB implementation. AFAIK the memory database implementation was from scratch, so that would benefit from some test coverage even though it looks good.

karalabe commented:
@Matthalp Added the memory iterator test. PTAL

fjl commented on database/iterator.go:
```diff
@@ -0,0 +1,61 @@
+}
+
+// Iteratorer wraps the NewIterator methods of a backing data store.
+type Iteratorer interface {
```
Suggestion: Iteratee

fjl commented:
More suggestions:

Memory database implementation could live inside package ethdb alongside the interface. This is possible because it has no dependencies.

`LevelDBDatabase` should be `LevelDB`.

kenny3486 commented on core/rawdb/database.go:
```diff
@@ -0,0 +1,51 @@
+	"github.com/ethereum/go-ethereum/internal/keyvalue"
+)
+
+// NewDatabase creates a high level database on top of a given key-value data
```
Hey boss, i read your comment from the Ethereum updates and was instantly very intrigued on your suggestions and extensively broad knowledge in the crypto space as a whole. I would like to speak to you regarding a business proposal if you are willing. Please email me at kenny3486@gmail.com and i look forward to speaking, thanks

Matthalp replied:
How would this interplay with #19200?

karalabe commented:
I've squashed and rebased everything since things got borked after the les merges. Fixed the conflicts. I also fixed all the issues raised in your first comment @fjl, as well as most raised in the second. We have a separate `leveldb` and `memorydb` packages within `ethdb` now.

In `rawdb` I left the `database.go` file reference both for now, because the freezer will mess things up anyway so I'd rather "fix" the dependency issues in that PR rather than dream up something abstract that will just brek with the freezer anyway.

PTAL

Matthalp commented on core/chain_indexer.go:
```diff
@@ -97,7 +97,7 @@ type ChainIndexer struct {
 // chain segments of a given size after certain number of confirmations passed.
 // The throttling parameter might be used to prevent database thrashing.
-func NewChainIndexer(chainDb, indexDb ethdb.Database, backend ChainIndexerBackend, section, confirm uint64, throttling time.Duration, kind string) *ChainIndexer {
+func NewChainIndexer(chainDb ethdb.Database, indexDb ethdb.Database, backend ChainIndexerBackend, section, confirm uint64, throttling time.Duration, kind string) *ChainIndexer {
```
Nit: consider moving this back to `chainDb, indexDb ethdb.Database`.

karalabe replied:
Meh :) I'm not much a fan of those shortened syntaxes :P

Matthalp commented on core/state/iterator_test.go:
```diff
@@ -51,12 +51,15 @@ func TestNodeIteratorCoverage(t *testing.T) {
 		}
 	}
-	for _, key := range db.TrieDB().DiskDB().(*ethdb.MemDatabase).Keys() {
+	it := db.TrieDB().DiskDB().(ethdb.Database).NewIterator()
```
I hate to add a last minute suggestion, but if this is a common pattern, it may be worth considering an `Database.forEach(func(key, value []byte)error) error` method that encapsulates the iteration and release:

```
func (db *Database) forEach(f func(key, value []byte)error) error {
  it := db.NewIterator()
  defer it.Release()
  for it.Next() {
    if err := f(it.Key(), it.Value()); err != nil {
      return err
    }
  }
  return nil
}
```

karalabe replied:
I think I'd rather avoid this for now at least. I'm more of a fan of explicit code and one way of doing things. Lets see if this appears a few more times in the database rewrites and extensions, and if it makes sense we can always add it then.

karalabe commented:
I'll just merge this PR to keep the ball rolling on the followup database reworks. We can fix any issues in future PRs, I think we all agreed that the general gist is OK now. Adding the freezer will also highlight anything we might have missed, or anything not needed, so it will be easier to discuss explicitly there than abstractly here.
